jupyterlab_gitsync/src/components/toolbar_buttons.tsx

```diff
@@ -0,0 +1,169 @@
+import * as React from 'react';
+import { classes, style } from 'typestyle';
+import {
+  toolbarItemClass,
+  toolbarButtonClass,
+  toolbarButtonIconClass,
+} from '../style/toolbar';
+
+import { Props } from './panel';
+
+export interface State {
+  state: string;
+  label: string;
+  style: ReturnType<typeof style>;
+}
+
+class ToolbarButton extends React.Component<Props, State> {
+  constructor(
+    props: Props,
+    init_label: string,
+    init_style: ReturnType<typeof style>
+  ) {
+    super(props);
+    this.state = {
+      state: 'Running',
+      label: init_label,
+      style: init_style,
+    };
+  }
+
+  render(): React.ReactElement {
+    return (
+      <div
+        className={classes(
+          'jp-ToolbarButton',
+          'jp-Toolbar-item',
+          toolbarItemClass
+        )}
+      >
+        <button
+          className={classes(
+            'bp3-button',
+            'bp3-minimal',
+            'jp-ToolbarButtonComponent',
+            'minimal',
+            'jp-Button',
+            toolbarButtonClass
+          )}
+          title={this.state.label}
+          onClick={this._onClick}
+        >
+          <span className={classes('bp3-button-text')}>
+            <span
+              className={classes(
+                'jp-ToolbarButtonComponent-icon',
+                toolbarButtonIconClass,
+                this.state.style
+              )}
+            >
+              {' '}
+            </span>
+          </span>
+        </button>
+      </div>
+    );
+  }
+
+  protected _onClick = (): void => {
+    console.log('hi');
+  };
+}
+
+export class ControlButton extends ToolbarButton {
```

Comment:
In general, I've not seen many inheritance patterns with React components and I've also not seen additional constructor parameters beyond props. React's official docs favor [composition](https://reactjs.org/docs/composition-vs-inheritance.html) over inheritance whereby you wrap the base component with additional functionality, state handling, etc. as opposed to creating a more tightly bound inheritance tree.

Reply:
This will be changed to follow React convention in a future PR.